### DISCUSSION TITLE
merge files with separator page

### DISCUSSION THREAD

**Discussion by ostiwe** [pdfcpu/pdfcpu#597] [General]
Hi, please tell me why it is impossible to merge several pdf files without using the file system?

I mean the api.

For example, I have 10 pdf files that need to be merged into one. At the same time, divide each of them with one empty sheet.

I saw the `before` parameter in the `InsertPagesFile` method, but it doesn't work as expected. I expect that after each "file" there will be a blank sheet. And as a result, a pdf is generated where an empty sheet is generated through each sheet.

It is also unclear how to work with the "selectedPages" parameter. I do not know how many pages there will be in each of the files.

Returning to the first question, why can't I collect `[]byte` of all files and then merge them? At the same time, add empty pages.

For example, I see it like this:

```go
	var fileFids = []string{
		"afd49fb5-3f73-4570-a01f-edb70a5142f5",
		"a78e3ec1-3d21-48b4-ba51-a67a9b0be02e",
		"855f0215-fe92-4bd9-b99b-4f650daaaa85",
	}

	var fileContents [][]byte

	for i := range fileFids {
		fileContent, err := storage.GetPDFFromExternalStorage(fileFids[i])
		if err != nil {
			panic(err)
		}

		fileContents = append(fileContents, fileContent)
	}


	pdfFile := pdfcpu.NewPDFFile("/some/file/name.pdf")

	for i := range fileContents {
		pdfFile.Append(fileContents[i])
		pdfFile.InsertBlankPage()
	}


	err := pdfFile.Save()
	if err != nil {
		panic(err)
	}
```

**Comment by hhrutter** [comment]
I wish it was that simple..
Merging PDFs really means merging two or more object based PDF cross reference tables respectively object models.
That's why you need to have these structs in memory.
In addition you need to analyze them (the xref tables) in order to process merging other PDF features like annotations, forms etc.

As far as including separator pages between your merged files I can think of 2 ways for you to achieve this right off the top of my head:

1) Create a PDF containing an empty page (similar to testdata/empty.pdf) then use it in your merge sequence:
`pdfcpu merge out.pdf in1.pdf empty.pdf in2.pdf empty.pdf in3.pdf`

2) Append an empty page to each of files before you merge them together:
You can use the `l`selector for last page:
`pdfcpu page insert -pages l -mode after in.pdf out.pdf`

.. or you can 🔧   hack right into the merge.go func Merge and add the empty page over there.
You are on your own if you go that route. 

PS: What works on the CLI like mentioned above works similarly using the API.

**Reply by ostiwe** [reply]
After a while I found another way.

I  used test code to create empty file from create_test.go, after deleting some lines of code.

It worked as I wanted.

If you added the same method to create an empty file, that would be great

```go
func CreateEmptyFile(savePath string) error {
	mediaBox := types.RectForFormat("A4")
	p := pdfcpuModel.Page{MediaBox: mediaBox, Fm: pdfcpuModel.FontMap{}, Buf: new(bytes.Buffer)}

	xRefTable, err := pdf.CreateDemoXRef()
	if err != nil {
		return err
	}

	rootDict, err := xRefTable.Catalog()
	if err != nil {
		return err
	}

	if err = pdf.AddPageTreeWithSamplePage(xRefTable, rootDict, p); err != nil {
		return err
	}

	if err = api.CreatePDFFile(xRefTable, savePath, nil); err != nil {
		return err
	}

	return nil
}
```

**Reply by hhrutter** [reply]
I guess this is one way to achieve what you want.

**Comment by hhrutter** [comment]
Closing this as stale.
